aspects/aspects.go

```diff
@@ -306,57 +328,144 @@ func (a *Aspect) Set(databag DataBag, request string, value interface{}) error {
 		return a.bundle.schema.Validate(data)
 	}
 
-	return &NotFoundError{
-		Account:    a.bundle.Account,
-		BundleName: a.bundle.Name,
-		Aspect:     a.Name,
-		Field:      request,
-		Cause:      "field not found",
+	return notFoundErrorFrom(a, request, "no matching rule")
+}
+
+// Get returns the aspect value identified by the request. If either the named
+// aspect or the corresponding value can't be found, a NotFoundError is returned.
+func (a *Aspect) Get(databag DataBag, request string, value *interface{}) error {
+	matches, err := a.matchGetRequest(request)
+	if err != nil {
+		return err
+	}
+
+	var merged interface{}
+	for _, match := range matches {
+		var val interface{}
+		if err := databag.Get(match.storagePath, &val); err != nil {
+			if errors.Is(err, PathNotFoundError("")) {
+				continue
+			}
+			return err
+		}
+
+		// build namespace around result
+		namespace := match.namespace
+		for i := len(namespace) - 1; i >= 0; i-- {
+			val = map[string]interface{}{namespace[i]: val}
+		}
+
+		// merge result with results from other matching rules
+		merged, err = merge(merged, val)
+		if err != nil {
+			return err
+		}
+	}
+
+	if merged == nil {
+		return notFoundErrorFrom(a, request, "no stored value for matching rules")
+	}
+
+	// the top level maps the request to the remaining namespace
+	*value = map[string]interface{}{request: merged}
+	return nil
+}
+
+func merge(old, new interface{}) (interface{}, error) {
+	if old == nil {
+		return new, nil
+	} else if new == nil {
+		return old, nil
+	}
+
+	oldType, newType := reflect.TypeOf(old).Kind(), reflect.TypeOf(new).Kind()
+	if oldType != newType {
+		return nil, fmt.Errorf("cannot merge results of different types %T, %T", old, new)
+	}
+
+	if oldType != reflect.Map {
+		// if the values are both scalars/lists, the new replaces the old value
+		return new, nil
+	}
+
+	// if the values are maps, merge them recursively
+	oldMap, newMap := old.(map[string]interface{}), new.(map[string]interface{})
+	for k, v := range newMap {
+		if storeVal, ok := oldMap[k]; ok {
+			merged, err := merge(storeVal, v)
+			if err != nil {
+				return nil, err
+			}
+			v = merged
+		}
+
+		oldMap[k] = v
 	}
+
+	return oldMap, nil
 }
 
-// Get returns the aspect value identified by the name. If either the named aspect
-// or the corresponding value can't be found, a NotFoundError is returned.
-func (a *Aspect) Get(databag DataBag, request string, value interface{}) error {
+type match struct {
+	// storagePath contains the storage path specified in the matching entry with
+	// any placeholders provided by the request filled in.
+	storagePath string
+
+	// namespace contains the unmatched parts of the entry's request.
```

Comment:
s/unmatched parts/rest nested suffix/ ?

the issue I have with "unmatched parts" is that there is no clear order sense to the term

Reply:
You're right, that's more indicative of what it is. There's a few of these, I'll replace them all with similar things. Thanks